### PR TITLE
Generate schema.json from Go struct

### LINKED ISSUE #56: Migrate schema.json to be generated from Go struct
**Is this a BUG REPORT or FEATURE REQUEST?**:
feature

**What happened**:
Currently we use schema.json as single source of truth.

**What you expected to happen**:
Use go struct as single source of truth and generate schema.json from it.

### PR DESCRIPTION
**Pull request by erbesharat**
**What this PR does / why we need it**:
Implements a new package called ``schema`` that is used for generating the schema.json file and validations.
  

**Which issue this PR fixes**:
fixes #56

### CODE REVIEW

File: schema/schema.go [hunk 1]
```diff
@@ -0,0 +1,694 @@
+package schema
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+)
+
+type Schema struct {
```
Comment (stp-ip) [inline schema/schema.go]: Couldn't we split this up into sub structs for easier usability and readability?
Just a thought on first view.

Comment (erbesharat) [reply schema/schema.go]: I think at first step it's a good idea to split each main property (like Core, Work) into a single struct.
Also, we can make something like ``StandardField`` struct for those fields that only have ``name`` and ``description`` inside them. We can also put an omitempty field inside this struct and use it for those structs that have ``format`` field too.

Comment (stp-ip) [reply schema/schema.go]: Yeah that sounds like a good plan.

File: cmd/resumic/main.go [hunk 2]
```diff
@@ -1,14 +1,51 @@
...
 func main() {
 	var doc string
+	var schemaFile string
 	flag.StringVar(&doc, "doc", "../../examples/invalid/invalid_email.json", "Example file")
+	flag.StringVar(&schemaFile, "schema", "./schema.json", "Generate JSON Schema")
 	flag.Parse()
-	validate.ValidateJSON(doc)
+
+	// Verify that a subcommand has been provided
+	if len(flag.Args()) < 1 {
```
Comment (stp-ip) [inline cmd/resumic/main.go]: Great to check this here.

**Comment by erbesharat** [comment]
@stp-ip Should we merge this? Arman is going to try using go tags to generate the schema.json and example.json in another PR.

**Comment by stp-ip** [comment]
Yeah happy to merge and then iterate. Also did a rough review, but as Arman also did a full one. Happy to merge now so you both can iterate.

File: cmd/resumic/main.go [hunk 3]
```diff
@@ -1,14 +1,51 @@
...
+		fmt.Println("subcommand is required")
+		os.Exit(1)
+	}
+	switch flag.Args()[0] {
+	case "validate":
+		validate.ValidateJSON(doc)
+	case "schema":
```
[the remark this answers is not in the document]
Comment (stp-ip) [reply cmd/resumic/main.go]: I would say let's iterate first and see, what we need or want and then split it out after we have something stable enough. Sounds good?

File: cmd/resumic/main.go [hunk 4]
```diff
@@ -1,14 +1,51 @@
...
 
 	"github.com/resumic/schema/cmd/resumic/validate"
+	"github.com/resumic/schema/schema"
 )
 
 func main() {
 	var doc string
+	var schemaFile string
 	flag.StringVar(&doc, "doc", "../../examples/invalid/invalid_email.json", "Example file")
+	flag.StringVar(&schemaFile, "schema", "./schema.json", "Generate JSON Schema")
```
[the remark this answers is not in the document]
Comment (stp-ip) [reply cmd/resumic/main.go]: @erbesharat Missed that comment before merge. Happy to take that in follow on PR, if we all agree flag.String is better. I don't have any particular preference, but can see the value in one less line.

**Pull request merged**